replica/table.cc

```diff
@@ -2041,12 +2044,6 @@ future<db::replay_position> table::discard_sstables(db_clock::time_point truncat
 
             cg.set_main_sstables(std::move(pruned));
             cg.set_maintenance_sstables(std::move(maintenance_pruned));
```

Comment:
The only exception unsafety seems to be in allocation, no? If so, then we can pre-allocate the on-stak vectors and capture them into invalidate()'s lambda

Reply:
the preparation must be done for all compaction groups, and on execution you apply the changes to all compaction groups, knowing that, on exception, you'll not leave it in partial state. row_cache::invalidate() accepts row_cache::external_updater that has prepare() and execute() methods, that we'll end up implementing for fixing the exception safety.

something like:
```
struct pruner : row_cache::external_updater {
    future<> prepare();   
    void execute() noexcept;
};
```

@bhalevy actually has an ongoing series touching this area, so I'd like to have his opinion.

Reply:
Even preallocation doesn't help, what are you going to do if preallocation fails?

The only realistic way to avoid it is to have an intrusive list with links in the sstable itself.

Reply:
Also, in practice, while we can avoid allocations with preallocation, we can't avoid allocations for the coroutine frame, and lots of allocations down seastar while removing the sstables. But at least those failures crash the system rather than leaving it in an inconsistent state.

Reply:
I thought Raphael's concern was about cache invalidation part only, wasn't it? Making prepare part throw-less will just extend this safety down to the place where sstables are atomically deleted, and the latter can still fail if, e.g. deletion log creation fails